server/block/campfire.go

```diff
@@ -0,0 +1,263 @@
+	}
+	if !c.Extinguished {
+		// if the campfire is water logged we extinguish it
+		if _, ok := w.Liquid(pos); ok {
+			c.Extinguished = true
+			w.PlaySound(pos.Vec3Centre(), sound.FireExtinguish{})
+			w.SetBlock(pos, c, nil)
+		}
+		for i, it := range c.Items {
+			if !it.Item.Empty() && it.Time <= 0 {
+				itemCooked := it.Item
+				if food, ok := itemCooked.Item().(item.Smeltable); ok {
+					ent := entity.NewItem(food.SmeltInfo().Product, pos.Vec3Middle())
+					ent.SetVelocity(mgl64.Vec3{rand.Float64()*0.2 - 0.1, 0.2, rand.Float64()*0.2 - 0.1})
+					w.AddEntity(ent)
+					c.Items[i].Item = item.Stack{}
+					w.SetBlock(pos, c, nil)
+					continue
+				}
+			}
+			c.Items[i].Time = it.Time - 1
+			w.SetBlock(pos, c, nil)
+		}
+	}
+}
+
+// EntityInside ...
+func (c Campfire) EntityInside(pos cube.Pos, w *world.World, e world.Entity) {
+	if flammable, ok := e.(entity.Flammable); ok {
+		// Try to egnite the campfire is the entity is on fire and ontop
+		if flammable.OnFireDuration() > 0 && c.Extinguished {
+			c.Extinguished = false
+			w.PlaySound(pos.Vec3(), sound.Ignite{})
+			w.SetBlock(pos, c, nil)
+		}
+		if !c.Extinguished && !w.RainingAt(pos) {
+			if l, ok := e.(entity.Living); ok && !l.AttackImmune() {
+				l.Hurt(c.Type.Damage(), damage.SourceFire{Campfire: true})
+			}
+			if flammable.OnFireDuration() < time.Second*8 {
+				flammable.SetOnFire(8 * time.Second)
+			}
+		}
+	}
+}
+
+// EncodeNBT ...
+func (c Campfire) EncodeNBT() map[string]any {
+	m := map[string]any{
+		"id": "Campfire",
+	}
+	for i, v := range c.Items {
+		itemNumberIdentifier := strconv.Itoa(i + 1)
+		if !v.Item.Empty() {
+			m["Item"+itemNumberIdentifier] = nbtconv.WriteItem(v.Item, true)
+			m["ItemTime"+itemNumberIdentifier] = uint8(v.Time)
+		}
+	}
+	switch c.Type {
+	case NormalFire():
+		m["id"] = "Campfire"
+		return m
+	case SoulFire():
+		m["id"] = "SoulFire"
+		return m
+	}
+	panic("invalid fire type")
+}
+
+// DecodeNBT ...
+func (c Campfire) DecodeNBT(data map[string]any) any {
+	for i := 0; i < 4; i++ {
+		itemNumberIdentifier := strconv.Itoa(i + 1)
+		c.Items[i] = CampfireItem{
+			Item: nbtconv.MapItem(data, "Item"+itemNumberIdentifier),
+			Time: int(nbtconv.Map[byte](data, "ItemTime"+itemNumberIdentifier)),
+		}
+	}
+	return c
+}
+
+// EncodeItem ...
+func (c Campfire) EncodeItem() (name string, meta int16) {
+	switch c.Type {
+	case NormalFire():
+		return "minecraft:campfire", 0
+	case SoulFire():
+		return "minecraft:soul_campfire", 0
+	}
+	panic("invalid fire type")
+}
+
+// EncodeBlock ...
+func (c Campfire) EncodeBlock() (name string, properties map[string]any) {
+	switch c.Type {
+	case NormalFire():
+		name = "minecraft:campfire"
+	case SoulFire():
+		name = "minecraft:soul_campfire"
+	default:
+		panic("invalid fire type")
+	}
+	return name, map[string]any{
+		"direction":    int32(horizontalDirection(c.Facing)),
+		"extinguished": c.Extinguished,
+	}
+}
+
+func allCampfires() (campfires []world.Block) {
```

Comment:
missing a doc here

Reply:
unexported functions do not require documentations